gene/insert.go

```diff
@@ -0,0 +1,326 @@
+
+			newProtSeq = frameshiftTranslate(g.codingSeq.seq[(codingPos+1)-frame:], g.utrThree.seq)
+
+			if newProtSeq[len(newProtSeq)-1] == dna.Stop {
+				answer.StopDist = len(newProtSeq) - 1
+			} else {
+				answer.StopDist = -2
+			}
+
+			var j int
+			for j = 0; newProtSeq[j] == g.protSeq[answer.AaPos]; j++ { // trim matching amino acids from frameshift
+				answer.AaPos++
+				if answer.StopDist != -2 {
+					answer.StopDist--
+				}
+				if answer.AaPos >= len(g.protSeq) {
+					break
+				}
+			}
+
+			answer.AaRef = []dna.AminoAcid{g.protSeq[answer.AaPos]}
+			answer.AaAlt = []dna.AminoAcid{newProtSeq[j]}
+
+			g.protSeq = newProtSeq //TODO append new prot seq to existing seq at point of insertion
+
+		} else { // In-Frame
+			newProtSeq = dna.TranslateToTer(g.codingSeq.seq) //TODO this can be much more efficient
+			answer.Consequence = InFrameInsertion
+			if frame != 0 { // Disrupts an existing codon
+				answer.AaAlt = dna.TranslateSeq(g.codingSeq.seq[(codingPos+1)-frame : (codingPos+1)+len(alt)+(3-frame)])
+				if answer.AaRef[0] == answer.AaAlt[0] {
+					answer.AaRef = nil
+					answer.AaAlt = answer.AaAlt[1:]
+					answer.AaPos++
+				}
+			} else { // Does not disrupt existing codons AND is inframe
+				answer.AaAlt = dna.TranslateSeq(alt)
+			}
+		}
+		g.protSeq = newProtSeq
+	} else { // NonCoding
+		// Update featureArray for noncoding
+		// figure out how the inserted bases should be labelled in the feature array. Priority: 5'UTR=3'Utr>Intron
+		fillVal := Feature(numbers.Min(int(g.featureArray[genomeIndexPos]), int(g.featureArray[genomeIndexPos+1])))
+
+		switch fillVal {
+		case UtrFive:
+			utrFiveOffset := 0
+
+			for i := 0; i <= genomeIndexPos; i++ {
+				if g.featureArray[i] == UtrFive {
+					utrFiveOffset++
+				}
+				if i > len(g.featureArray) { // case of no codingSeq
+					break
+				}
+			}
+
+			if utrFiveOffset > -1 { // Update 5' UTR
+				err = insertAdjust(g, &g.cdnaSeq, g.utrFive.start+utrFiveOffset, alt)
+				if err != nil {
+					return EffectPrediction{}, err
+				}
+			}
+
+		case UtrThree:
+			utrThreeOffset := 0
+
+			for i := genomeIndexPos; g.featureArray[i] < 0; i-- {
+				if g.featureArray[i] == UtrThree {
+					utrThreeOffset++
+				}
+				if i <= 0 { // case of no codingSeq
+					break
+				}
+			}
+
+			if utrThreeOffset > -1 { // Update 3' UTR
+				err = insertAdjust(g, &g.cdnaSeq, g.utrThree.start+utrThreeOffset, alt)
+				if err != nil {
+					return EffectPrediction{}, err
+				}
+			}
+		}
+
+		tmpSlice := make([]Feature, len(alt))
+		g.featureArray = append(g.featureArray, tmpSlice...)
+		copy(g.featureArray[genomeIndexPos+1+len(alt):], g.featureArray[genomeIndexPos+1:])
+		for i := 0; i < len(alt); i++ {
+			g.featureArray[genomeIndexPos+1+i] = fillVal
+		}
+
+		var endCdnaOffset int
+		answer.CdnaPos, answer.CdnaDist, err = GenomicPosToCdna(g, genomePos+1)
+		_, endCdnaOffset, err = GenomicPosToCdna(g, genomePos+1+(len(alt)-1))
+		if err != nil { //TODO REMOVE
```

Comment:
I am not sure why this error check would be removed.

Reply:
removed the todo